OpenRA.Mods.Common/Traits/World/EditorActionManager.cs

```diff
@@ -0,0 +1,165 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class EditorActionManagerInfo : TraitInfo<EditorActionManager> { }
+
+	public class EditorActionManager : IWorldLoaded
+	{
+		Stack<EditorActionContainer> undoStack = new Stack<EditorActionContainer>();
+		Stack<EditorActionContainer> redoStack = new Stack<EditorActionContainer>();
+
+		public event Action<EditorActionContainer> ItemAdded;
+		public event Action<EditorActionContainer> ItemRemoved;
+
+		int nextId;
+
+		public void WorldLoaded(World w, WorldRenderer wr)
+		{
+			Add(new OpenMapAction());
+		}
+
+		public void Add(IEditorAction editorAction)
+		{
+			editorAction.Execute();
+
+			foreach (var action in undoStack)
+				action.Status = EditorActionStatus.History;
+
+			var actionContainer = new EditorActionContainer(nextId++, editorAction);
+
+			ClearRedo();
+			undoStack.Push(actionContainer);
+
+			if (ItemAdded != null)
+				ItemAdded(actionContainer);
+		}
+
+		public void Undo()
+		{
+			if (undoStack.Count == 1)
+				return;
+
+			var editorAction = undoStack.Pop();
+			undoStack.Peek().Status = EditorActionStatus.Active;
+			editorAction.Action.Undo();
+			editorAction.Status = EditorActionStatus.Future;
+			redoStack.Push(editorAction);
+		}
+
+		void ClearRedo()
+		{
+			if (!HasRedos())
+				return;
+
+			while (redoStack.Count != 0)
+			{
+				var item = redoStack.Pop();
+
+				if (ItemRemoved != null)
+					ItemRemoved(item);
+			}
+		}
+
+		public void Redo()
+		{
+			if (redoStack.Count == 0)
+				return;
+
+			var editorAction = redoStack.Pop();
+
+			editorAction.Status = EditorActionStatus.Active;
+			editorAction.Action.Do();
+			undoStack.Peek().Status = EditorActionStatus.History;
+			undoStack.Push(editorAction);
+		}
+
+		public bool HasUndos()
+		{
+			return undoStack.Count > 1;
+		}
+
+		public bool HasRedos()
+		{
+			return redoStack.Count > 0;
+		}
+
+		public void Rewind(int id)
+		{
+			while (undoStack.Peek().Id != id)
+				Undo();
+		}
+
+		public void Forward(int id)
+		{
+			while (undoStack.Peek().Id != id)
+				Redo();
+		}
+	}
+
+	public enum EditorActionStatus
+	{
+		History,
+		Active,
+		Future,
+	}
+
+	public interface IEditorAction
+	{
+		void Execute();
+		void Do();
+		void Undo();
+
+		string Text { get; }
```

Comment:
For usability, I think it might be nice if the action text contained some minimal context, e.g. "Actor added: 3tnk", "Copied 20 tiles". This would obviously need to be minimal so as to fit in the label width.

Possibly, it might be worth adding a tooltip that gave more details of the change on hover.

Not required, just throwing the idea out there.

Reply:
Good suggestion!

Reply:
Because this will be called every render, we should cache the string to reduce allocations. Consider making every implementation of the Text property `public string Text { get; private set; }` and setting the value in the `Do` method.

(Sorry - should've realised this when suggesting it originally!)